<a href="https://colab.research.google.com/github/Shinevenbrighter/Implementacion-patrones-regex/blob/main/Actividad_2_2_Implementaci%C3%B3n_de_Patrones_Regex.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Tecnológico de Monterrey Campus Toluca

Actividad 2.2 Implementación de Patrones Regex

Alumna: Valentina Tejeda Fuentes A01771768

Materia: Análisis y diseño de algoritmos avanzados

Grupo: 601

Profesor: Marcial Roberto Leyva Fernandez

##Introducción

"Las expresiones regulares, también conocidas como **regex**, son un conjunto de patrones utilizados para buscar, editar y manipular texto y datos. El término se originó en la utilidad UNIX ed. Las regex tienen varios usos, incluida la validación de la entrada del usuario, la búsqueda de archivos y cadenas, y la sustitución de texto dentro de una cadena." (¿Qué Es Regex Y Cómo Usarlo? | Lenovo México, 2021)

En esta actividad se desarrolla un tipo "formulario" que permite validar diferentes tipos de datos personales mediante expresiones regulares (regex). El objetivo principal es garantizar que la información ingresada cumpla con un formato correcto antes de ser aceptada, simulando así un sistema de captura de datos seguro y estructurado.





##Programa de validación

El formulario solicita de manera secuencial los siguientes datos: nombre, apellidos, edad, fecha de nacimiento, teléfono, dirección, correo electrónico y CURP. Cada dato se verifica automáticamente con un patrón de validación específico, asegurando que se cumplan reglas de sintaxis y consistencia, por ejemplo:
- Que los nombres y apellidos inicien con mayúscula.

- Que la edad esté dentro de un rango lógico (1 a 120 años).

- Que el teléfono y correo electrónico tengan un formato válido.

- Que el CURP cumpla con la estructura oficial y sea consistente con el nombre y la fecha de nacimiento de la persona.

In [ ]:
%%writefile formulario.py
import re

# -------------------------------
# Funciones validadoras
# -------------------------------
def checknombre(my_str):
    patron = r"^[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(\s[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)*$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

def checkedad(my_str):
    patron = r"^(?:[1-9][0-9]?|1[01][0-9]|120)$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

def checkphone(my_str):
    patron = r"^\+?\d{1,3}?\s?\d{7,15}$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

def checkdireccion(my_str):
    patron = r"^[\w\s\.,#\-]+$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

def checkemail(my_str):
    patron = r"^[\w\.-]+@[\w\.-]+\.\w{2,}$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

def checkcurp(my_str):
    patron = r"^[A-Z]{4}\d{6}[HM][A-Z]{5}[A-Z0-9]\d$"
    res = bool(re.fullmatch(patron, my_str))
    print(f'{my_str} "{patron}" -> {"Coincide" if res else "No coincide"}')
    return res

# -------------------------------
# Generar CURP base esperado
# -------------------------------
def generar_curp(nombre, apellidos, fecha, sexo="H", estado="DF"):
    apellido_paterno, apellido_materno = apellidos.split()
    nombre = nombre.strip().upper()

    curp = (
        apellido_paterno[0].upper() +
        re.sub(r"[^AEIOU]", "", apellido_paterno[1:].upper())[0] +
        apellido_materno[0].upper() +
        nombre[0].upper()
    )

    yyyy, mm, dd = fecha.split("-")
    curp += yyyy[2:] + mm + dd
    curp += sexo
    curp += estado

    def primera_consonante(palabra):
        cons = re.sub(r"[AEIOU]", "", palabra.upper()[1:])
        return cons[0] if cons else "X"

    curp += (
        primera_consonante(apellido_paterno) +
        primera_consonante(apellido_materno) +
        primera_consonante(nombre)
    )

    curp += "0A1"
    return curp

# -------------------------------
# Función para ejecutar formulario
# -------------------------------
def ejecutar_formulario():
    print("=== Formulario de Registro ===\n")

    # Nombre
    while True:
        nombre = input("★ Nombre(s): ")
        if checknombre(nombre):
            print()
            break
        print("Nombre inválido, intenta de nuevo.\n")

    # Apellidos
    while True:
        apellidos = input("★ Apellidos (Paterno Materno): ")
        validos = True
        for x in apellidos.split():
            if not checknombre(x):
                validos = False
        if validos:
            print()
            break
        print("Apellidos inválidos, intenta de nuevo.\n")

    # Edad
    while True:
        edad = input("★ Edad: ")
        if checkedad(edad):
            print()
            break
        print("Edad inválida, intenta de nuevo.\n")

    # Fecha de nacimiento
    patron_fecha = r"^\d{4}-\d{2}-\d{2}$"
    while True:
        fecha = input("★ Fecha de nacimiento (YYYY-MM-DD): ")
        if re.fullmatch(patron_fecha, fecha):
            print(f'{fecha} "{patron_fecha}" -> Coincide\n')
            break
        print("Formato de fecha inválido (ejemplo: 2000-05-12), intenta de nuevo.\n")

    # Teléfono
    while True:
        tel = input("★ Teléfono (Ejemplo: +52 5561084772): ")
        if checkphone(tel):
            print()
            break
        print("Teléfono inválido, intenta de nuevo.\n")

    # Dirección
    while True:
        direccion = input("★ Dirección: ")
        if checkdireccion(direccion):
            print()
            break
        print("Dirección inválida, intenta de nuevo.\n")

    # Email
    while True:
        email = input("★ Email: ")
        if checkemail(email):
            print()
            break
        print("Email inválido, intenta de nuevo.\n")

    # CURP
    curp_esperado = generar_curp(nombre, apellidos, fecha, sexo="M")
    print(f"(Referencia CURP esperada según tus datos: {curp_esperado})\n")

    while True:
        curp = input("★ CURP: ").upper()

        if not checkcurp(curp):
            print("CURP inválido, intenta de nuevo.\n")
            continue

        if curp.startswith(curp_esperado[:10]):
            break
        else:
            print("CURP válido pero NO coincide con nombre y fecha, intenta de nuevo.\n")

    print("\n === Todos los patrones pasaron, el registro ha sido completado ===\n")


# -------------------------------
# PROGRAMA PRINCIPAL
# -------------------------------
for i in range(1, 5):
    print(f"\n>>> Caso de prueba {i} <<<")
    ejecutar_formulario()


Overwriting formulario.py


In [ ]:
!python formulario.py


>>> Caso de prueba 1 <<<
=== Formulario de Registro ===

★ Nombre(s): Salvador
Salvador "^[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(\s[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)*$" -> Coincide

★ Apellidos (Paterno Materno): Alvarado Reyes
Alvarado "^[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(\s[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)*$" -> Coincide
Reyes "^[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(\s[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)*$" -> Coincide

★ Edad: 20
20 "^(?:[1-9][0-9]?|1[01][0-9]|120)$" -> Coincide

★ Fecha de nacimiento (YYYY-MM-DD): 2004-11-10
2004-11-10 "^\d{4}-\d{2}-\d{2}$" -> Coincide

★ Teléfono (Ejemplo: +52 5561084772): 5568735192
5568735192 "^\+?\d{1,3}?\s?\d{7,15}$" -> Coincide

★ Dirección: Chalco 306 Col sector popular Toluca
Chalco 306 Col sector popular Toluca "^[\w\s\.,#\-]+$" -> Coincide

★ Email: tootalvarey@gmail.com
tootalvarey@gmail.com "^[\w\.-]+@[\w\.-]+\.\w{2,}$" -> Coincide

(Referencia CURP esperada según tus datos: AARS041110MDFLYL0A1)

★ CURP: AARS041110HMCLYLA2
AARS041110HMCLYLA2 "^[A-Z]{4}\d{6}[HM][A-Z]{5}[A-Z0-9]\d$" -> Coincide

 === Todos 

##Resultados

Durante la ejecución del programa se realizaron distintos registros con datos personales simulados. Cada uno de los registros pasó por las validaciones implementadas con expresiones regulares, verificando campos como nombre, edad, dirección, correo electrónico, teléfono, CURP y fecha de nacimiento.

En todos los experimentos se observó que los patrones fueron reconocidos correctamente, lo cual confirma que las expresiones regulares utilizadas funcionan de manera adecuada para validar la información de entrada.

| Experimento | Nombre                           | Edad | Dirección                                           | Correo                         | Telefono   | CURP               | Fecha nacimiento | Resultados |
|-------------|----------------------------------|------|-----------------------------------------------------|--------------------------------|------------|--------------------|------------------|------------|
| 1           | Salvador Alvarado Reyes        | 20  | Chalco 306 Col sector popular Toluca | tootalvarey@gmail.com | 5568735192 | AARS041110HMCLYLA2 | 2004-11-10         | Todos los patrones pasaron, el registro ha sido completado           |
| 2           | Brisa Sofía León Pérez           | 19   | Mariano Arista 701 bellavista metepec               | brisasofia02@gmail.com         | 7223029987 | LEPB050930MPLNRRA4 | 2005-09-30         |Todos los patrones pasaron, el registro ha sido completado            |
| 3           | Valentina Tejeda Fuentes         | 21   | Villas del campo 1295 calimaya estado de mexico     | valen.2604@hotmail.com         | 5561084772 | TEFV040426MHGJNLA1 | 2004-04-26         |  Todos los patrones pasaron, el registro ha sido completado          |
| 4           | Mariana Isabel Aguilar Espindola | 20   | Privada lazaro cardenas colonia san juan buenavista | mariana.agesp@gmail.com        | 7221755988 | AUEM050211MMCGSRA1 | 2005-02-11         |  Todos los patrones pasaron, el registro ha sido completado          |

##Conclusión

La implementación de expresiones regulares or regex es una herramienta eficaz para validar información en formularios de registro. A lo largo de los cuatro casos de prueba se comprobó que los patrones diseñados permiten identificar correctamente datos como nombres, edades, direcciones, correos electrónicos, teléfonos y CURP, reduciendo posibles errores en la captura de información.

Igualmente, se da una retroalimentación inmediata al usuario al solicitar nuevamente los datos cuando no cumplían con el formato esperado, lo que asegura  que la información guardada en un registro sea confiable y sobre todo consistente. También se realiza una detección de inconsistencias relacionadas con la identidad de los participantes, como la correspondencia del CURP con el nombre y la fecha de nacimiento.

Finalmente, el uso de regex en este tipo de aplicaciones representa una solución práctica y robusta para la validación de datos en sistemas que requieren entradas precisas y estructuradas, garantizando mayor integridad en la información recolectada

##Referencias


¿Qué es Regex y cómo usarlo? | Lenovo México. (2021). Lenovo.com. https://www.lenovo.com/mx/es/glosario/expresion-regular-regex/?orgRef=https%253A%252F%252Fwww.google.com%252F&srsltid=AfmBOoqnmxfcvXgHUzK0Kg_9GzT4BHQUEFyRbwvNzX3XWCu277nGo895

‌